### ISSUE TITLE
analyze.quantification errors

### ISSUE THREAD

**Issue by vagarwal87** [YosefLab/MPRAnalyze#6]
Hi, I finally got all of my data prepared, setup the MPRAobject, and completed estimateDepthFactors. During the analyze.quantification step, however, I'm getting many errors that look like this:

> Fitting model...
> error fitting: A:HNF4A-ChMod_chr10:11917871-11917984_[chr10:11917842-11918013]:015: Error in optim(par = guess, fn = cost.dnarna, llfnDNA = ll.funs$dna, llfnRNA = ll.funs$rna, : L-BFGS-B needs finite values of 'fn'
> 
> error fitting: A:HNF4A-ChMod_chr10:34165653-34165745_[chr10:34165613-34165784]:076: Error in optim(par = guess, fn = cost.dnarna, llfnDNA = ll.funs$dna, llfnRNA = ll.funs$rna, : L-BFGS-B needs finite values of 'fn'
> 
> error fitting: A:HNF4A-ChMod_chr10:60797400-60797480_[chr10:60797354-60797525]:026: Error in optim(par = guess, fn = cost.dnarna, llfnDNA = ll.funs$dna, llfnRNA = ll.funs$rna, : L-BFGS-B needs finite values of 'fn'

Finally, the program terminates like this:

> Error in models[valids][[1]] : subscript out of bounds
> Calls: analyze.quantification -> reformatModels -> extractProp
> Execution halted

Any idea what is going on or how to fix this? Happy to send some sample data to reproduce the error if it helps.

Thanks!
Vikram

**Comment by talashuach** [comment]
Hi Vikram,
It seems like for some reason the likelihood function is giving non-finite values, so the optimization function fails. The final error indicates that this apparently happened with all of the inserts.

I haven't been able to reproduce this, but if you can send me some sample data I'll be happy to look into what's going on.

Thanks!
Tal

**Comment by vagarwal87** [comment]
Ok, to reproduce the issue, here is the code I'm using:

```
library(MPRAnalyze)
load("testMpraObject.Rdata")
obj <- estimateDepthFactors(obj, lib.factor = c("repbatch", "condition"), which.lib = "both", depth.estimator = "uq") #can change to "totsum" or other normalizations
obj <- analyze.quantification(obj = obj, dnaDesign = ~ barcode + repbatch + condition, rnaDesign = ~ condition)

```

And here is the sample dataset (gunzip first):
[testMpraObject.Rdata.gz](https://github.com/YosefLab/MPRAnalyze/files/2117524/testMpraObject.Rdata.gz)

**Comment by talashuach** [comment]
Hi Vikram,
Thanks!

Your annotations data.frame has both "repbatch" and "barcode" as numeric variables, instead of factors. So the model treats them as continuous covariates instead of categorical ones.
This shouldn't crash the model, so I'm still looking into that. But it's probably not what you want, and if you set both covariates to factors the model runs fine.

Do note however that your data is huge, which is great for statistical power, but also means that the model has > 800 coefficients to optimize, and that might take a while to run.
I'm working on speeding up the runtime regardless, but the current version runs somewhat slow on data of this size. I highly recommend parallelizing and running it on a cluster if you have one available.

Parallelization in MPRAnalyze is done with the bioconductor BiocParallel package (An intro to using it can be found [here](http://bioconductor.org/packages/release/bioc/vignettes/BiocParallel/inst/doc/Introduction_To_BiocParallel.pdf)). There you can find clear explanations on how to register different backends, depending on the machine and resources you have available.
Basically the easiest way to do it is simply:
```
library(BiocParallel)
par.backend <- bpparam()
obj <- MpraObject(<same arguments as before>, BPPARAM = par.backend)
```

Thanks for helping me locate these issues!
Tal

**Comment by vagarwal87** [comment]
Hi Tal,

Thanks for the helpful tips! Great point, I'll definitely convert them into factors then, had forgotten they'd be treated as numerics. Yes I was actually using BiocParallel but removed that code to reproduce the error in a simple way. Will definitely use the parallelized version for the final dataset. Thanks for looking into why it crashes in the meantime!

Best,
Vikram

**Comment by arushiv** [comment]
Hi,

Was this issue resolved? I get similar errors using my data if I use estimateDepthFactors before calling analyze.quantification. The code finishes without errors if I don't call estimateDepthFactors.

**Comment by talashuach** [comment]
This issue was indeed resolved, but it sounds like your issue is a different one. If you have an abundance of 0 observations in your data, the upper-quartile depth calculation could return 0s as "library sizes", which crashes the analysis.

You can solve this by installing the newest version of the package, which warns about this and tries avoiding this issue by computing the size factors only based on the non-0 observations:
`install.packages("devtools")
devtools::install_github("YosefLab/MPRAnalyze")`

Alternatively, you can solve this by using the `totsum` computation which uses the total sum of counts instead of the quantiles, and doesn't result in 0 factors.

**Comment by arushiv** [comment]
Hi Tal,

Thanks for the quick reply. Yeah, I realized that with my data if I used the "uq" option for RNA depth normalization, the 'rnaDepth' slot was all 0s. I was going to type a fuller description of the issue. 'totsum' does get around the issue so thanks for suggesting that. I'll have to think more about the differences between using totsum or quantiles in my case. I do see that you suggest 'uq' method in general, so I'll try using the new version.

Thanks for your help!

PS: I remembered your poster from Biology of Genomes last year and had searched for this tool. Reading the preprint was very helpful!

**Comment by talashuach** [comment]
Quick correction: the fix to the UQ computation is only included in the version I just pushed a few moments ago (1.1.6), so if you already updated - please do so again. :)

As for which estimation to use - using quantiles is generally more robust to outlying values, which is why I recommend it, but if you don't have too many outliers either one should work fine.

I'm glad to hear you remembered the poster! Please let me know if you have any more questions\issues with using the package.

Tal
